src/cast_core/src/ExternalScriptRunner.cpp

```diff
@@ -66,29 +67,72 @@ struct ExternalScriptRunner::PhaseConfig {
 std::string ExternalScriptRunner::exec(const char* cmd) {
     std::array<char, 128> buffer;
     std::string result;
-    std::unique_ptr<FILE, decltype(&pclose)> pipe(popen(cmd, "r"), pclose);
+
+    FILE* pipe = popen(cmd, "r");
     if (!pipe) {
         throw std::runtime_error("Execution of command " + std::string(cmd) + " failed!");
     }
-    while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr) {
-        result += buffer.data();
+    try {
+        while (fgets(buffer.data(), buffer.size(), pipe) != nullptr) {
+            result += buffer.data();
+            BOOST_LOG_TRIVIAL(info) << "Script output: " << buffer.data();
+        }
     }
+    catch (...) {
+        pclose(pipe);
+        throw;
+    }
+
+    int pcloseResult = pclose(pipe);
+    int exitStatus = WEXITSTATUS(pcloseResult);
+    if(exitStatus != 0) {
+        throw std::runtime_error("Script exited with non-zero exit code " + std::to_string(exitStatus));
+    }
+
     return result;
 }
 
+class TempScriptFile {
+public:
+    TempScriptFile(const std::string& script) {
+        boost::filesystem::path temp = boost::filesystem::temp_directory_path() / boost::filesystem::unique_path();
+        _name = temp.native();
+        _file = std::ofstream(_name);
+        _file << script;
+
+        // File must be flushed to ensure the contents are available when
+        // the command is invoked
+        _file.flush();
+    }
+
+    TempScriptFile(const TempScriptFile&) = delete;
+
+    ~TempScriptFile() {
+        std::remove(_name.c_str());
+    }
+
+    const std::string& name() {
+        return _name;
+    }
+private:
+    std::ofstream _file;
+    std::string _name;
+};
+
 void ExternalScriptRunner::run() {
     // This section gets executed before the phases. Setup is executed here, so maybe
     // we could add some sanity checks before actually running the shell.
 
     for (auto&& config : _loop) {
-
-        BOOST_LOG_TRIVIAL(debug) << "ExternalScriptRunner running " << config->programCommand;
-
         for (const auto&& _ : config) {
+            TempScriptFile file{config->script};
+            std::stringstream fullInvocation;
+            fullInvocation << config->invocation << " "<< file.name() << " 2>&1";
+            std::string invocation = fullInvocation.str();
 
             // Execute the script and read result from stdout
             auto ctx = config->scriptOperation.start();
```

Comment:
Not your change, but it might be worth a comment to say that line 134 starts the timer on the metrics, but does not actually start the script. I got confused by that.

Reply:
Updated

Reply:
Thank you!